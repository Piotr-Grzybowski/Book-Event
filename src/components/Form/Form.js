import React from 'react';
import DatePicker from 'react-datepicker';
import Message from '../Message/Message';
import "./react-datepicker.css";
import PropTypes from 'prop-types';

function Form({ bookingData, setValue, sendBooking, message}) {
  // workaround for datePicker inside issue
  const formatDate = (date) => {
    return new Date(date.getTime() - (date.getTimezoneOffset() * 60000));
  }
  return (
    <div>
      <form data-testid="form" onSubmit={(event) => { event.preventDefault(); sendBooking(bookingData)}} >
        <input data-testid="firstName" type="text" value={bookingData.firstName} required onChange={(event) => { setValue({ firstName: event.target.value }) }}></input>
        <input data-testid="lastName" type="text" value={bookingData.lastName} required onChange={(event) => { setValue({ lastName: event.target.value }) }}></input>
        <input data-testid="email" type="email" value={bookingData.email} required onChange={(event) => { setValue({ email: event.target.value }) }}></input>
        <DatePicker selected={bookingData.date} value={bookingData.date} onChange={(date) => setValue({ date: formatDate(date) })} required />
        <button data-testid="submit">Submit</button>
      </form>
      <Message message={message} loader={bookingData.sending} />
    </div>
  );
}

Form.propTypes = {
  bookingData: PropTypes.object,
  setValue: PropTypes.func,
  sendBooking: PropTypes.func,
  getMessage: PropTypes.object
}

export default Form;