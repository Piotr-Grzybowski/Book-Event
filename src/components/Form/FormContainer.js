import { connect } from 'react-redux';
import Form from './Form';
import { setValue, sendBooking, getBooking, getMessage} from '../../redux/bookingRedux';

const mapStateToProps = (state) => ({
  bookingData: getBooking(state),
  message: getMessage(state)
});

const mapDispatchToProps = (dispatch) => ({
  sendBooking: (data) => dispatch(sendBooking(data)),
  setValue: (value) => dispatch(setValue(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(Form);