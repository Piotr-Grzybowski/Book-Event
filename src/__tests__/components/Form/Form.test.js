import React from 'react';
import Form from '../../../components/Form/Form';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import register from 'ignore-styles';
register(['.css', '.sass', '.scss']);

describe('Form component test:', () => {
  // workaround for date-picker issue
  const formatDate = (date) => {
    return new Date(date.getTime() - (date.getTimezoneOffset() * 60000));
  }
  const props = {
    bookingData: {
      firstName: 'John',
      lastName: 'Doe',
      email: 'example@example',
      date: new Date(1811112121119),
      sending: {
        active: false,
        error: false
      }
    },
    message: ' ',
    setValue: jest.fn(),
    sendBooking: jest.fn()
  }
  let getByTestId, container, firstName, lastName, email, date, submitButton;
  const setValue = jest.fn();
  const sendBooking = jest.fn();
  
  beforeEach(() => {
    ({ container, getByTestId } = render(<Form {...props} setValue={setValue} sendBooking={sendBooking} />));
    firstName = getByTestId('firstName');
    lastName = getByTestId('lastName');
    email = getByTestId('email');
    date = container.querySelector('.react-datepicker__input-container').firstChild;
    submitButton = getByTestId('submit');
  });
  afterEach(() => {
    cleanup();
  });
  
  it('should match snapshot', () => {
    const snapshot = renderer.create(<Form {...props} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });
  it('should render form with correct inputs', () => {
    const form = getByTestId('form');
    expect(form).toBeInTheDocument();
    expect(firstName).toBeInTheDocument();
    expect(lastName).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(date).toBeInTheDocument();
  });
  it('should value in input be the same like in props', () => {
    expect(firstName.value).toBe(props.bookingData.firstName);
    expect(lastName.value).toBe(props.bookingData.lastName);
    expect(email.value).toBe(props.bookingData.email);
    expect(new Date(date.value).toLocaleDateString('en-US')).toBe(formatDate(props.bookingData.date).toLocaleDateString('en-US'));
  })
  it('should call setValue method when input value changes', () => {
    fireEvent.change(firstName, { target: { value: 'Chuck' }});
    expect(setValue).toHaveBeenCalledWith({firstName: 'Chuck'});
    fireEvent.change(lastName, { target: { value: 'Johnson' }});
    expect(setValue).toHaveBeenCalledWith({lastName: 'Johnson'});
    fireEvent.change(email, { target: { value: 'Chuck@somewhere' }});
    expect(setValue).toHaveBeenCalledWith({email: 'Chuck@somewhere'});
    fireEvent.change(date, { target: { value: new Date(411112121119) }})
    expect(setValue).toHaveBeenCalledTimes(4);
  })
  it('should call sendBooking method on submit with input values', () => {
    fireEvent.submit(submitButton);
    expect(sendBooking).toHaveBeenCalledTimes(1);
    expect(sendBooking).toHaveBeenCalledWith({
      firstName: firstName.value,
      lastName: lastName.value,
      email: email.value,
      date: new Date(1811112121119),
      sending: {
        active: false,
        error: false
      }
    });
  })
})