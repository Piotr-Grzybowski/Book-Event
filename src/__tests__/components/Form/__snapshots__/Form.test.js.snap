// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component test: should match snapshot 1`] = `
<div>
  <form
    data-testid="form"
    onSubmit={[Function]}
  >
    <input
      data-testid="firstName"
      onChange={[Function]}
      required={true}
      type="text"
      value="John"
    />
    <input
      data-testid="lastName"
      onChange={[Function]}
      required={true}
      type="text"
      value="Doe"
    />
    <input
      data-testid="email"
      onChange={[Function]}
      required={true}
      type="email"
      value="example@example"
    />
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          className=""
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          required={true}
          type="text"
          value="05/24/2027"
        />
      </div>
    </div>
    <button
      data-testid="submit"
    >
      Submit
    </button>
  </form>
  <div
    className="message"
    data-testid="message"
  >
    <p>
       
    </p>
  </div>
</div>
`;
